ml-prims/src/hmm/magma/b_likelihood.h

```diff
@@ -0,0 +1,242 @@
+
+        for (size_t obsId = j_start; obsId < nObs; obsId+=nThreads_y) {
+                for (size_t clId = i_start; clId < nCl; clId+=nThreads_x) {
+
+                        size_t idx = IDX(clId, obsId, nCl);
+                        dX_batches[idx] = dX_array[obsId];
+                        dmu_batches[idx] = dmu_array[clId];
+                        dInvSigma_batches[idx] = dInvSigma_array[clId];
+                }
+        }
+}
+
+template <typename T>
+void create_llhd_batches(int nObs, int nCl,
+                         T **&dX_batches, T **&dmu_batches, T **&dInvSigma_batches,
+                         T **&dX_array, T **&dmu_array, T **&dInvSigma_array){
+        dim3 block(32, 32, 1);
+        dim3 grid(ceildiv((int) nCl, (int) block.x),
+                  ceildiv((int) nObs, (int) block.y),
+                  1);
+
+        int nThreads_x = grid.x * block.x;
+        int nThreads_y = grid.y * block.y;
+
+        createLlhdBatchesKernel<T> <<< grid, block >>>(nObs, nCl,
+                                                       dX_batches, dmu_batches, dInvSigma_batches,
+                                                       dX_array, dmu_array, dInvSigma_array,
+                                                       nThreads_x, nThreads_y);
+        cudaDeviceSynchronize();
+        CUDA_CHECK(cudaPeekAtLastError());
+}
+
+
+template <typename T>
+__global__
+void subtractBatchedKernel(magma_int_t m, magma_int_t n, magma_int_t batchCount,
+                           T** dO_array, magma_int_t lddO,
+                           T** dX_array, magma_int_t lddx,
+                           T** dY_array, magma_int_t lddy,
+                           int nThreads_x, int nThreads_y, int nThreads_z){
+
+        int i_start = threadIdx.x + blockDim.x * blockIdx.x;
+        int j_start = threadIdx.y + blockDim.y * blockIdx.y;
+        int k_start = threadIdx.z + blockDim.z * blockIdx.z;
+
+        int idxO, idxX, idxY;
+        // TODO : Check the difference
+
+        for (size_t bId = k_start; bId < batchCount; bId+=nThreads_z) {
+                for (size_t j = j_start; j < n; j+=nThreads_x) {
+                        for (size_t i = i_start; i < m; i+=nThreads_y) {
+                                idxO = IDX(i, j, lddO);
+                                idxX = IDX(i, j, lddx);
+                                idxY = IDX(i, 0, lddy);
+                                // printf("%f\n", (float) dX_array[bId][idxX] - dY_array[bId][idxY]);
+                                // printf("%d\n", idxO);
+                                dO_array[bId][idxO] = dX_array[bId][idxX] - dY_array[bId][idxY];
+                        }
+                }
+        }
+}
+
+template <typename T>
+void subtract_batched(magma_int_t m, magma_int_t n, magma_int_t batchCount,
+                      T** dO_array, magma_int_t lddO,
+                      T** dX_array, magma_int_t lddx,
+                      T** dY_array, magma_int_t lddy){
```

Comment:
accept stream argument

Reply:
Same the file is more relevant for the next GMM PR